### ISSUE TITLE
Integration of Royale SDK download through Moonshine App Store Helper v.2.2.0

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#451]
Requirements:

1. Royale SDK download through Moonshine App Store Helper v2.2.0
2. Royale SDK detection in Moonshine v1.16.0

**Comment by rat-moonshine** [comment]
Things to cover:

- [ ] Copyright label changes to the MASH inteface
- [ ] License Agreement addition for Royale
- [ ] Remove Apache FlexJS download
- [ ] Verify checksum for distributed Royale js-swf SDK
- [ ] Host the archives for Royale js-swf to moonshine-ide.com
- [ ] Ensure Royale SDK downloaded into ~/Downloads/MoonshineSDKs/Royale_SDK/ path
- [ ] Ensure Moonshine able detect the downloaded Royale SDK to above path in OSX properly
- [ ] Ensure detected and listed Royale SDK run properly in Moonshine

**Comment by piotrzarzycki21** [comment]
> * Host the archives of Royale js-swf to moonshine-ide.com

What's the point of storing Royale on our servers ? We should use Apache Mirrors.

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 , I have downloaded the js-swf Royale SDK from this link: https://royale.apache.org/download/. Extracted the archive and created a new Royale project in Moonshine. 

However, when ran with 'Build & Run as Javascript' this thrown me following error:

```
: Compiling NewJavaScriptBrowserProject
: SDK path: /Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs
: Using Royale Compiler codebase: /Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs
: Using Royale SDK: /Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs
: MXMLJSC
: +royalelib=/Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks
: -sdk-js-lib=/Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks/js/Royale/generated-sources
: -load-config+=obj/NewJavaScriptBrowserProjectConfig.xml
: -compiler.targets=JSRoyale
: -js-output=/Users/devsena/Desktop/NewJavaScriptBrowserProject
: /Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks/royale-config.xml(74): col: 0 Error: unable to open '/Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks/libs/player/11.1/playerglobal.swc'.
: /Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks/royale-config.xml (line: 74)
:       </external-library-path>
: /Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks/royale-config.xml(98): col: 0 Error: unable to open '/Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks/libs/player/11.1'.
: /Users/devsena/Downloads/MoonshineSDKs/Royale_SDK/apache-royale-0.9.4-bin-js-swf/royale-asjs/frameworks/royale-config.xml (line: 98)
:       </library-path>
: 1.51293756 seconds
```

Does these error tells you anything?

**Comment by piotrzarzycki21** [comment]
> Does these error tells you anything?

Do you have in environment setup AIR_HOME - It's probably needed in case of that SDK no matter if you build to SWF or JS - Not sure. Try to setup and make sure that is available, restart Moonshine and build again.

**Comment by rat-moonshine** [comment]
The error is a bit misleading if this reasoned by the AIR_HOME environment variable.

However, when I checked I found I do have AIR_HOME setup to my environment variable:
```
export AIR_HOME=/Users/devsena/ApacheFlexJSFrameworkSource/air-sdk
export PATH=$PATH:$AIR_HOME
```

Since we requires to run binaries from `~/Downloads` folder in sandbox, I also changed the AIR_HOME path to run from `~/Downloads`, but that didn't change anything.

Did you ever tried to run Royale project with js-swf Royale SDK on OSX before?

Sandbox version not able to access environment variable - but we need to be sure if the above error because of this field, then we would require to set alternative course to provide this dependency.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine Setup then [PLAYERGLOBAL_HOME](https://github.com/apache/royale-asjs/blob/develop/README.md#additional-prerequisites-for-swf-output)

**Comment by piotrzarzycki21** [comment]
> Did you ever tried to run Royale project with js-swf Royale SDK on OSX before?

No I didn't check that. Let me try and see what I have

**Comment by rat-moonshine** [comment]
I setup `PLAYERGLOBAL_HOME` on my Windows system to test:
![image](https://user-images.githubusercontent.com/24370870/52553596-03370980-2e0a-11e9-95f2-f78c2796aa98.png)

Upon restarting Moonshine I created a new Royale project, I see similar error when tried to build as JavaScript:

```
: G:\DevareaLocal\ApacheFlexSDK\apache-royale-0.9.4-bin-js-swf\royale-asjs\frameworks\royale-config.xml(74): col: 0 Error: unable to open 'G:\DevareaLocal\ApacheFlexSDK\apache-royale-0.9.4-bin-js-swf\royale-asjs\frameworks\libs\player\11.1\playerglobal.swc'.
: G:\DevareaLocal\ApacheFlexSDK\apache-royale-0.9.4-bin-js-swf\royale-asjs\frameworks\royale-config.xml (line: 74)
: 
:       </external-library-path>
```

I'm not sure despite of having `PLAYERGLOBAL_HOME` (which has different path), why it's continue to look inside Royale SDK `frameworks/libs` folder.

@piotrzarzycki21 can you give some tests if the above situation runs in Windows even?

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I will give it a test and let you know

**Comment by piotrzarzycki21** [comment]
I have tested stuff on Windows.

1. I have remove from my environment variables `AIR_HOME`, `FLEX_HOME `and `PLAYERGLOBAL_HOME`. Remember that it should be also removed from `Path`
2. I make sure that all those variables are not visible anymore by doing in cmd `echo %PLAYERGLOBAL_HOME%`
3. After that I tried build hello world project in Moonshine and got the same issue as you
4. I have added `PLAYERGLOBAL_HOME` to folder `d:\Work\player\` which contains folder `11.1` with `playerglobal.swc` file
5. I added `PLAYERGLOBAL_HOME` to `Path`
6. I run build and it successfully build hello world app.

**Comment by piotrzarzycki21** [comment]
Here another thing which is related to js-swf build and we shouldn't try to change that, cause it is related to framework.
If user would like to use js-swf version of Royale he need to take care of adding [playerglobal.swc](https://helpx.adobe.com/flash-player/kb/archived-flash-player-versions.html) on his own by doing:
1. Add environment variable `PLAYERGLOBAL_HOME`
2. Add playerglobal.swc file to downloaded SDK to folder `frameworks\libs\player\{playerglobalversion}\playerglobal.swc`

On Mac platform we cannot access environment variable from sandbox application, so there is only option 2 for those users. 

If user don't want to do above instructions - js only version avoid that. MASH may have bigger description for js-swf version with above instruction - That's my only suggestion. That description doesn't have to be added in this release.

**Comment by rat-moonshine** [comment]
> If user don't want to do above instructions - js only version avoid that. MASH may have bigger description for js-swf version with above instruction - That's my only suggestion. That description doesn't have to be added in this release.

Yes, I guess that is right. In future, when we have select option to choose between things, then we might give description on advantage and dependency differences, which may help the user to select between the two (js-only and js-swf). 

For the current version, since there is no select option, we have to distribute only any - js-only or js-swf.

**Comment by rat-moonshine** [comment]
After some discussion on this with Joel, we choose to go by js-only SDK for the v2.2.0 release. We probably want to re-visit the dependency part (PLAYERGLOBAL_HOME) in future version, though.

**Comment by JoelProminic** [comment]
> > * Host the archives of Royale js-swf to moonshine-ide.com
> 
> What's the point of storing Royale on our servers ? We should use Apache Mirrors.

@piotrzarzycki21

Previously, we have linked Moonshine App Store Helper downloads to specific Apache mirrors.  This caused trouble because the mirrors would stop working, breaking the application.  We started hosting the specific downloads on moonshine-ide.com to avoid this problem.

Alternatively, we could develop a system to:
1. Open the list of Apache mirrors and select an appropriate mirror (hopefully using the mirror suggested by Apache)
2. Download the appropriate version of the SDK from that mirror
3. Verify the checksum or signature of the download.

If you guys like this idea, we can create a separate issue to work on this.

**Comment by rat-moonshine** [comment]
> Alternatively, we could develop a system to:

I think that is something we can invest in to. I agree to this idea.

**Comment by piotrzarzycki21** [comment]
No matter what in my opinion we need to use Apache mirrors, cause right now I have no idea whether we are not violating any license by hosting that files on our servers.

Btw. what kind of problems did we have in the past? What was those mirrors, are we saying about Apache Flex SDK mirrors or previous FlexJS ?

**Comment by rat-moonshine** [comment]
As far I can remember in earlier days of Apache SDK Installer, we were seeing download/mirror problems at different times; Maybe, we had also seen the problem against Apache Ant. But we never host Flex/JS SDK in our server but installed by the Apache SDK Installer source only. Probably because of the same reason we choose to download Apache Ant from our server. Recently, Royale also.

If we can take a look a little more in the distribution agreements/requisites, and able to host relevant SDKs from our site, then that might solve many different problems, including one with 64-bit requirement. This will also let us reduce our current dependency over Apache SDK Installer source.

**Comment by piotrzarzycki21** [comment]
> As far I can remember in earlier days of Apache SDK Installer, we were seeing download/mirror problems at different times; Maybe, we had also seen the problem against Apache Ant. But we never host Flex/JS SDK in our server but installed by the Apache SDK Installer source only. Probably because of the same reason we choose to download Apache Ant from our server. Recently, Royale also.
> 
> If we can take a look a little more in the distribution agreements/requisites, and able to host relevant SDKs from our site, then that might solve many different problems, including one with 64-bit requirement. This will also let us reduce our current dependency over Apache SDK Installer source.

Are you 100% sure that you have seen problems with Apache Mirrors not other part of SDK? Apache Flex installer downloading only one file from Apache mirrors at the beginning. It has less than 100 MB and those are Flex binaries. If there is no MD5 mismatch issue - I personally have seen during past years only couple of failures in that area. Most of the time there were problem with downloading other parts.

**Comment by rat-moonshine** [comment]
No, I'm not sure standing at this time but I can only say we used to see different errors while downloading complete SDK package with its different components; I know it's Flex SDK that download as a single file, and therefore AIR SDK followed by different small components. I don't think I'm able to give you specific details of those errors now. 

Looking into the older issue documented, the most obvious one I found when we had trouble downloading Apache Ant through their provided mirror. However, we examined their license on distribution, based on standard Apache License unless we're not going to modify their binaries, restrictions not applied on distribution, so we hosted Apache Ant only.

Royale is also includes int he same Apache License 2.0 if I'm right.
